storage/testsuite/long_bench.go

```diff
@@ -0,0 +1,801 @@
+				pathData.Close())
+		}
+		kvi.closeFunc = func() error { return utils.CombineErrors(gzReader.Close(), pathData.Close()) }
+		reader = gzReader
+	} else {
+		kvi.closeFunc = pathData.Close
+	}
+	kvi.scanner = bufio.NewScanner(reader)
+	return kvi, nil
+}
+
+// Next should be called by BulkImporter instances in order to advance the iterator. It fills in
+// a storage.ListItem instance, and returns a boolean indicating whether to continue. When false is
+// returned, iteration should stop and nothing is expected to be changed in item.
+func (kvi *KVInputIterator) Next(item *storage.ListItem) bool {
+	if !kvi.scanner.Scan() {
+		kvi.reachedEnd = true
+		kvi.err = kvi.scanner.Err()
+		return false
+	}
+	if kvi.err != nil {
+		return false
+	}
+	kvi.itemNo++
+	parts := bytes.Split(kvi.scanner.Bytes(), []byte("\t"))
+	if len(parts) != 3 {
+		kvi.err = errs.New("Invalid data in %q on line %d: has %d fields", kvi.fileName, kvi.itemNo, len(parts))
+		return false
+	}
+	k, err := interpolateInput(parts[1])
+	if err != nil {
+		kvi.err = errs.New("Failed to read key data from %q on line %d: %v", kvi.fileName, kvi.itemNo, err)
+		return false
+	}
+	v, err := interpolateInput(parts[2])
+	if err != nil {
+		kvi.err = errs.New("Failed to read value data from %q on line %d: %v", kvi.fileName, kvi.itemNo, err)
+		return false
+	}
+	item.Key = storage.Key(k)
+	item.Value = storage.Value(v)
+	item.IsPrefix = false
+	return true
+}
+
+// Error() returns the last error encountered while iterating over the input file. This must be
+// checked after iteration completes, at least.
+func (kvi *KVInputIterator) Error() error {
+	return kvi.err
+}
+
+func openTestData(tb testing.TB) *KVInputIterator {
+	tb.Helper()
+	pathDataFile, err := pathToPathDataFile()
+	if err != nil {
+		tb.Fatalf("Failed to locate file with test data: %v", err)
+	}
+	inputIter, err := newKVInputIterator(pathDataFile)
+	if err != nil {
+		tb.Fatal(err)
+	}
+	return inputIter
+}
+
+// BulkImporter identifies KV storage facilities that can do bulk importing of items more
+// efficiently than inserting one-by-one.
+type BulkImporter interface {
+	BulkImport(storage.Iterator) error
+}
+
+// BulkCleaner identifies KV storage facilities that can delete all items efficiently.
+type BulkCleaner interface {
+	BulkDelete() error
+}
+
+// BenchmarkPathOperationsInLargeDb runs the "long benchmarks" suite for KeyValueStore instances.
+func BenchmarkPathOperationsInLargeDb(b *testing.B, store storage.KeyValueStore) {
+	if !*doLongBenchmarks {
+		b.Skip("Long benchmarks not enabled.")
+	}
+
+	initialized := false
+
+	// create these anew with each invocation of this function; we only want these to run "once"
+	// per storage backend under test.
+	var initOnce sync.Once
+	var cleanupOnce sync.Once
+
+	doTest := func(name string, testFunc func(*testing.B, storage.KeyValueStore)) {
+		b.Run(name, func(bb *testing.B) {
+			initOnce.Do(func() { initStore(bb, store, &initialized) })
```

Comment:
You don't need to use sync.Once things. It's sufficient if you do in the beginning of function:

```
if !*disableInit { initStore(b, store) }
if !*disableCleanup { defer cleanupStore(b, store) }
```

Reply:
I did it like this because I thought I had found a case where the import and cleanup were still being run without any tests, because of the filtering options I had applied, and I didn't want those to run unless something was actually going to make use of the data (it's pretty time-consuming). That might not be an issue worth worrying about, though. I think I have changed the structure of the tests pretty significantly since that point. If you're really worried about it, I'll take out the sync.Once stuff.

Reply:
This slow down would happen only when:
1. you have `--test-bench-long`
2. you have run `-bench PathOperationsInLargeDb/XXX`

Neither happens often, so I wouldn't worry about it and keep code simpler.